### PR TITLE
Proxy Kythe apis to codesearch server

### PR DESCRIPTION
**Pull request by tylerwilliams**
1) Build ts protos of kythe stuff we're going to use in the UI.
2) Add a util/grpc_forward library called from grpc_server.go that forwards traffic, only if forwarding is configured, to some backend server.

I considered just manually implementing these methods in buildbuddy_service and calling the kythe backend service (writing out the proxy basically), but there are between 10 and 24 methods we'd want to proxy to the kythe backend, so this seemed cleaner and simpler. It also allows us to proxy other stuff in the future (like the codesearch APIs, for example).

LMK what you think, I'm open.

### CODE REVIEW

File: server/util/grpc_forward/grpc_forward.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/buildbuddy-io/buildbuddy/server/util/flag"
+	"github.com/buildbuddy-io/buildbuddy/server/util/grpc_client"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"github.com/mwitkow/grpc-proxy/proxy"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+var (
+	kytheBackendToProxy = flag.String("app.kythe_backend_to_proxy", "", "If set, proxy kythe traffic to this backend")
```
Comment (bduffany) [inline server/util/grpc_forward/grpc_forward.go:17]: nit: the other proxy-ish flags are `-cache_proxy.remote_cache` and `-build_event_proxy.hosts` - maybe `-kythe_proxy.target` would be a good name for this?

**Review by bduffany — APPROVED** [review]

File: server/util/grpc_forward/grpc_forward.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+func init() {
+	mu.Lock()
+	backendConnectionPools = make(map[string]*grpc_client.ClientConnPool)
+	mu.Unlock()
+}
+
+type directorFunc func(ctx context.Context, fullMethodName string) (context.Context, *grpc.ClientConn, error)
+
+func getProxyDirector() directorFunc {
+	if *kytheBackendToProxy != "" {
```
Comment (vadimberezniker) [inline server/util/grpc_forward/grpc_forward.go:32]: if *kytheBackendToProxy == "" {
  return nil
}
the rest

File: server/util/grpc_forward/grpc_forward.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		return func(ctx context.Context, fullMethodName string) (context.Context, *grpc.ClientConn, error) {
+			target := ""
+			if strings.HasPrefix(fullMethodName, "/kythe.service") {
```
Comment (vadimberezniker) [inline server/util/grpc_forward/grpc_forward.go:35]: If this is meant to be a general forwarding util, there shouldn't be any kythe-specific references either here or in the flag name.

**Review by vadimberezniker — CHANGES_REQUESTED** [review]

File: server/util/grpc_forward/grpc_forward.go (hunk 1 above)
Comment (tylerwilliams) [reply server/util/grpc_forward/grpc_forward.go:17]: Ack (removed flag)

File: server/util/grpc_forward/grpc_forward.go (hunk 2 above)
Comment (tylerwilliams) [reply server/util/grpc_forward/grpc_forward.go:32]: Ack. (removed flag)

File: server/util/grpc_forward/grpc_forward.go (hunk 3 above)
Comment (tylerwilliams) [reply server/util/grpc_forward/grpc_forward.go:35]: Done.

File: server/util/grpc_forward/grpc_forward.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	once                   sync.Once
+	mu                     sync.RWMutex
+	backendConnectionPools map[string]*grpc_client.ClientConnPool
+)
+
+func lookupProxyTarget(fullMethodName string) (string, error) {
+	target := ""
+	for _, pair := range *proxyTargets {
+		if strings.HasPrefix(fullMethodName, pair.Prefix) {
+			target = pair.Target
```
Comment (vadimberezniker) [inline server/util/grpc_forward/grpc_forward.go:35]: Why not directly return here?

**Review by bduffany — APPROVED** [review]

**Review by vadimberezniker — APPROVED** [review]

File: server/util/grpc_forward/grpc_forward.go (hunk 4 above)
Comment (tylerwilliams) [reply server/util/grpc_forward/grpc_forward.go]: Done.

**Pull request merged**